File: QuickStarts/CS/EventBroker/Tests/Tests.EventBrokerExtension/EventBrokerWireupStrategyFixture.cs
//===============================================================================
// Microsoft patterns & practices
// Unity Application Block
//===============================================================================
// Copyright © Microsoft Corporation.  All rights reserved.
// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
// OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
// LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS FOR A PARTICULAR PURPOSE.
//===============================================================================

using System;
using System.Collections.Generic;
using EventBrokerExtension;
using Microsoft.Practices.ObjectBuilder2;
using Microsoft.Practices.Unity;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using SimpleEventBroker;
using Tests.EventBrokerExtension.TestObjects;
using Tests.EventBrokerExtension.Utility;

namespace Tests.EventBrokerExtension
{
    [TestClass]
    public class EventBrokerWireupStrategyFixture
    {
        [TestMethod]
        public void StrategyThrowsIfWireupIsNeededAndBrokerNotInLocator()
        {
            MockBuilderContext context = CreateContext();
            NamedTypeBuildKey buildKey = NamedTypeBuildKey.Make<OneEventPublisher>();
            EventBrokerInfoPolicy policy = new EventBrokerInfoPolicy();
            policy.AddPublication("paste", "Pasting");
            context.Policies.Set<IEventBrokerInfoPolicy>(policy, buildKey);

            try
            {
                OneEventPublisher existing = new OneEventPublisher();
                context.ExecuteBuildUp(buildKey, existing);
            }
            catch (Exception)
            {
                // If we got here, we're Ok, this is expected.
                return;
            }
            Assert.Fail("No exception occured");
        }

        [TestMethod]
        public void NoExceptionIfExistingObjectDoesntAndNoBroker()
        {
            MockBuilderContext context = CreateContext();
            NamedTypeBuildKey buildKey = NamedTypeBuildKey.Make<OneEventPublisher>();
            EventBrokerInfoPolicy policy = new EventBrokerInfoPolicy();
            policy.AddPublication("paste", "Pasting");
            context.Policies.Set<IEventBrokerInfoPolicy>(policy, buildKey);

            context.ExecuteBuildUp(buildKey, null);
            // No assert needed, if we got here, we're ok            
        }

        [TestMethod]
        public void ExceptionIfNoWireupNeededAndNoBroker()
        {
            MockBuilderContext context = CreateContext();
            NamedTypeBuildKey buildKey = NamedTypeBuildKey.Make<object>();
            EventBrokerInfoPolicy policy = new EventBrokerInfoPolicy();
            context.Policies.Set<IEventBrokerInfoPolicy>(policy, buildKey);

            try
            {
                context.ExecuteBuildUp(buildKey, new object());
            }
            catch (Exception)
            {
                // If we got here, we're ok            
                return;
            }
            Assert.Fail("No exception Occurred");
        }

        [TestMethod]
        public void StrategyProperlyWiresEvents()
        {
            MockBuilderContext context = CreateContext();
            NamedTypeBuildKey buildKey = NamedTypeBuildKey.Make<ClipboardManager>();

            EventBroker broker = new EventBroker();
            var brokerLifetime = new ExternallyControlledLifetimeManager();
            brokerLifetime.SetValue(broker);
            context.Policies.Set<ILifetimePolicy>(brokerLifetime, NamedTypeBuildKey.Make<EventBroker>());

            EventBrokerInfoPolicy policy = new EventBrokerInfoPolicy();
            policy.AddPublication("cut", "Cut");
            policy.AddPublication("copy", "Copy");
            policy.AddPublication("paste", "Paste");

            policy.AddSubscription("copy", typeof(ClipboardManager).GetMethod("OnCopy"));
            policy.AddSubscription("clipboard data available",
                                   typeof(ClipboardManager).GetMethod("OnClipboardDataAvailable"));

            context.Policies.Set<IEventBrokerInfoPolicy>(policy, buildKey);

            ClipboardManager existing = new ClipboardManager();

            context.ExecuteBuildUp(buildKey, existing);

            List<string> registeredEvents = new List<string>(broker.RegisteredEvents);
            registeredEvents.Sort();

            List<string> expectedEvents = new List<string>(new string[]
            {
                "cut",
                "copy",
                "paste",
                "clipboard data available"
            });
            expectedEvents.Sort();

            CollectionAssert.AreEqual(expectedEvents, registeredEvents);
        }

        private MockBuilderContext CreateContext()
        {
            MockBuilderContext context = new MockBuilderContext();
            context.Strategies.Add(new LifetimeStrategy());
            context.Strategies.Add(new EventBrokerWireupStrategy());

            return context;
        }
    }
}

File: quickstarts/VB/EventBroker/Tests/Tests.EventBrokerExtension/EventBrokerWireupStrategyFixture.vb
'===============================================================================
' Microsoft patterns & practices
' Unity Application Block
'===============================================================================
' Copyright © Microsoft Corporation.  All rights reserved.
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
' OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
' LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
' FITNESS FOR A PARTICULAR PURPOSE.
'===============================================================================

Imports System
Imports System.Text
Imports System.Collections.Generic
Imports EventBrokerExtension
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports Microsoft.Practices.ObjectBuilder2
Imports SimpleEventBroker
Imports Tests.EventBrokerExtension.Tests.EventBrokerExtension.TestObjects
Imports Tests.EventBrokerExtension.Tests.EventBrokerExtension.Utility
Imports Microsoft.Practices.Unity

<TestClass()> Public Class EventBrokerWireupStrategyFixture

    <TestMethod(), ExpectedException(GetType(InvalidOperationException))> _
    Public Sub StrategyThrowsIfWireupIsNeededAndBrokerNotInLocator()
        Dim context As MockBuilderContext = CreateContext()
        Dim buildKey As New NamedTypeBuildKey(GetType(OneEventPublisher))
        Dim policy As New EventBrokerInfoPolicy()
        policy.AddPublication("paste", "Pasting")
        context.Policies.Set(Of IEventBrokerInfoPolicy)(policy, buildKey)

        Dim existing As New OneEventPublisher()
        context.ExecuteBuildUp(buildKey, existing)
    End Sub

    <TestMethod()> _
    Public Sub NoExceptionIfExistingObjectDoesntAndNoBroker()
        Dim context As MockBuilderContext = CreateContext()
        Dim buildKey As New NamedTypeBuildKey(GetType(OneEventPublisher))
        Dim policy As New EventBrokerInfoPolicy()
        policy.AddPublication("paste", "Pasting")
        context.Policies.Set(Of IEventBrokerInfoPolicy)(policy, buildKey)

        context.ExecuteBuildUp(buildKey, Nothing)
        ' No assert needed, if we got here, we're ok
    End Sub

    <TestMethod(), ExpectedException(GetType(InvalidOperationException))> _
    Public Sub ExceptionIfNoWireupNeededAndNoBroker()
        Dim context As MockBuilderContext = CreateContext()
        Dim buildKey As New NamedTypeBuildKey(GetType(OneEventPublisher))
        Dim policy As New EventBrokerInfoPolicy()
        context.Policies.Set(Of IEventBrokerInfoPolicy)(policy, buildKey)

        context.ExecuteBuildUp(buildKey, New Object())

    End Sub

    <TestMethod()> _
    Public Sub StrategyProperlyWiresEvents()
        Dim context As MockBuilderContext = CreateContext()
        Dim buildKey As New NamedTypeBuildKey(GetType(ClipboardManager))

        Dim broker As New EventBroker()
        Dim brokerLifetime As New ExternallyControlledLifetimeManager
        brokerLifetime.SetValue(broker)
        context.Policies.Set(Of ILifetimePolicy)(brokerLifetime, NamedTypeBuildKey.Make(Of EventBroker)())

        Dim policy As New EventBrokerInfoPolicy
        policy.AddPublication("cut", "Cut")
        policy.AddPublication("copy", "Copy")
        policy.AddPublication("paste", "Paste")

        policy.AddSubscription("copy", GetType(ClipboardManager).GetMethod("OnCopy"))
        policy.AddSubscription("clipboard data available", _
                               GetType(ClipboardManager).GetMethod("OnClipboardDataAvailable"))

        context.Policies.Set(Of IEventBrokerInfoPolicy)(policy, buildKey)

        Dim existing As New ClipboardManager()

        context.ExecuteBuildUp(buildKey, existing)

        Dim registeredEvents As New List(Of String)(broker.RegisteredEvents)
        registeredEvents.Sort()

        Dim expectedEvents As New List(Of String)(New String() { _
            "cut", _
            "copy", _
            "paste", _
            "clipboard data available" _
        })

        expectedEvents.Sort()

        CollectionAssert.AreEqual(expectedEvents, registeredEvents)
    End Sub

    Private Function CreateContext() As MockBuilderContext
        Dim context As New MockBuilderContext()
        context.Strategies.Add(New LifetimeStrategy())
        context.Strategies.Add(New EventBrokerWireupStrategy())
        Return context
    End Function
End Class
